header/js/typeahead-new.js

```diff
@@ -0,0 +1,221 @@
+/*global fetch*/
+const Delegate = require('ftdomdelegate');
+const debounce = require('../../utils').debounce;
+import { SuggestionList } from './suggestion-list';
+
+const React = require('react');
+const ReactDom = require('react-dom');
+
+
+function getNonMatcher (container) {
+	if (typeof container === 'string') {
+		return function (el) {
+			return el && el !== document && !el.matches(container);
+		};
+	}
+
+	return function (el) {
+		return el && el !== document && el !== container;
+	};
+}
+
+function isOutside (el, container) {
+	const doesntMatch = getNonMatcher(container);
+
+	while (doesntMatch(el)) {
+		el = el.parentNode;
+	}
+
+	return !el || el === document;
+}
+
+function regExpEscape (s) {
+	return s.replace(/[-\\^$*+?.()|[\]{}]/g, '\\$&');
+};
+
+class Typeahead {
+	constructor (containerEl, dataSrc, showAllHandler) {
+		this.container = containerEl;
+		this.searchEl = this.container.querySelector('input[type="text"]');
+		this.submitButton = this.container.querySelector('button[type="submit"]')
+		this.showAllHandler = showAllHandler;
+		this.dataSrc = dataSrc;
+		this.minLength = 2;
+		this.showAllItem = false;
+		this.init();
+	}
+
+	init () {
+		this.suggestions = [];
+		this.suggestionListContainer = document.createElement('div');
+		this.container.insertBefore(this.suggestionListContainer, this.submitButton);
+		this.suggestionsView = ReactDom.render(<SuggestionList/>, this.suggestionListContainer);
+
+		if (this.showAllItem) {
+			this.viewAllItem = document.createElement('li');
+			this.viewAllItem.classList.add('o-header__typeahead-view-all');
+			this.viewAllItemInnerHTML = '<button type="submit" data-trackable="view-all">View All Results</button>';
+		}
+
+		this.delegate = new Delegate(this.container);
+		this.bodyDelegate = new Delegate(document.body);
+		this.suggest = this.suggest.bind(this);
+		this.onType = debounce(this.onType, 150).bind(this);
+		this.onDownArrow = this.onDownArrow.bind(this);
+		this.onSuggestionKey = this.onSuggestionKey.bind(this);
+		this.onSuggestionClick = this.onSuggestionClick.bind(this);
```

Comment:
I don't know react that well, and would refactoring bits of the component that don't actually need to repaint benefit the user anyway?